src/metabase/api/search.clj

```diff
@@ -386,16 +379,18 @@
                                          :when (seq query)]
                                      query)}
           _            (log/tracef "Searching with query:\n%s" (u/pprint-to-str search-query))
-          ;; sort results by [model name]
-          results      (sort-by (juxt (comp model->sort-position :model)
-                                      :name)
-                                (db/query search-query :max-rows search-max-results))]
-      (for [row results
-            :when (check-permissions-for-model row)]
-        ;; MySQL returns `:favorite` and `:archived` as `1` or `0` so convert those to boolean as needed
-        (-> row
-            (update :favorite bit->boolean)
-            (update :archived bit->boolean))))))
+          results      (db/query search-query :max-rows search-max-results)]
+      ;; sort by [score model name]
+      (sort-by (juxt
+                (comp - (partial search/score (:search-string search-ctx)))
```

Comment:
the pattern `(sort-by - ...)` has some pitfalls with floats: https://clojure.org/guides/comparators#_beware_using_subtraction_in_a_comparator

tldr the output of compare must be a integer. if you return a float, it is cast and this can lead to loss of resolution:
```
(sort - [1.0 0.9 0.8 0.7]) #_ -> (1.0 0.9 0.8 0.7)
(sort - [1 2 3 4]) #_ -> (1 2 3 4)
```

Reply:
TIL, thanks!

Unless I'm misunderstanding, using the unary `-` as the `keyfn` will invoke `compare`, which uses traditional `<`, `==`, and `>` to arrive at `-1`, `0`, and `1`. This code uses rationals, but things seem good for both rationals and floats:

```
(sort-by - [1/2 1/3 2/3 4/3 10/3])
;; => (10/3 4/3 2/3 1/2 1/3)
(sort-by - [0.1 0.2 0.4 0.3 0.9 0.15 1.5])
;; => (1.5 0.9 0.4 0.3 0.2 0.15 0.1)
```


```
(defn sort-by
  "Returns a sorted sequence of the items in coll, where the sort
  order is determined by comparing (keyfn item).  If no comparator is
  supplied, uses compare.  comparator must implement
  java.util.Comparator.  Guaranteed to be stable: equal elements will
  not be reordered.  If coll is a Java array, it will be modified.  To
  avoid this, sort a copy of the array."
  {:added "1.0"
   :static true}
  ([keyfn coll]
   (sort-by keyfn compare coll))
  ([keyfn ^java.util.Comparator comp coll]
   (sort (fn [x y] (. comp (compare (keyfn x) (keyfn y)))) coll)))
   ```

Reply:
oh right. good point about the compare versus keyfn.